modules/compiler/utils/source/work_item_loops_pass.cpp

```diff
@@ -1343,6 +1356,27 @@ void setUpLiveVarsAlloca(compiler::utils::BarrierWithLiveVars &barrier,
   }
 }
 
+// Returns true if the original barrier call happens in a basic block terminated
+// by a call to llvm.trap followed by an unreachable instruction.
+bool isBarrierInTrapBB(const compiler::utils::BarrierWithLiveVars &barrierMain,
+                       unsigned id) {
+  if (auto *const CI = barrierMain.getBarrierCall(id)) {
+    BasicBlock *originalBB = CI->getParent();
+    Instruction *terminator = originalBB->getTerminator();
+    if (isa<UnreachableInst>(terminator) &&
+        originalBB->sizeWithoutDebug() >= 2) {
+      Instruction &pre_term = *(++originalBB->rbegin());
+      if (auto *ci = dyn_cast<CallInst>(&pre_term)) {
+        auto called = ci->getCalledFunction();
+        if (called->getName() == "llvm.trap") {
```

Comment:
Does this work if there are multiple blocks or is there guaranteed to be only one here?

I'm wondering whether we need to check for the trap since if we hit an unreachable as our only successor we are unable to continue anyway.

Reply:
If there are multiple blocks then the barrier region will have more than zero successors I think, so it should go in the code path that handles that case, I'll add more test cases

Reply:
is that true? 
If you have block 1 branch unconditional to  block 2(terminate), isn't that just single block terminate?

Or if we have more than one unreachable terminates they wouldn't be in the first block.

I don't know this code that well, so I could just be misunderstanding - it should basically follow the same logic as num_sucessors when it gets built